Text/us-499 Documentation to address Sigma Level Output (#1105)
NOAA-EMC/UPP: doc/UserGuide/CustomizingTheUPP/EnablingOutput.rst

      RUN | control_p8                                        | - noaacloud                          | baseline |
      RUN | regional_control                                  |                                      | baseline |

#. Save and exit: Type ``esc`` and ``:wq``. Then hit ``Enter``/``Return``.

#. Edit each test's output frequency by modifying the test files under ``ufs-weather-model/tests/tests``. For example, to output every three hours for 24 hours:

   .. code-block:: console

      export OUTPUT_FH='3 -1'

   .. note:: 

      ``'3 -1'`` means every 3 hours. The ``-1`` indicates that it is not a list of hours. For example, someone could choose to output an arbitrary list of hours like ``export OUTPUT_FH='0 4 5 7 12 16 19'``. 

6. If you are not running on NOAA :term:`RDHPCS` or do not have access to the ``stmp*`` disk space on those systems, alter the ``dprefix`` path for the machine you are running on in ``rt.sh`` to point to a directory (such as ``$TOP_DIR``) where you have write permissions. For example:

   .. code-block:: console

      dprefix=${dprefix:-"path/to/${TOP_DIR}"}

   .. note::

>       For users not running on NOAA :term:`RDHPCS`, see the UFS WM documentation for :ref:`building the Weather Model on other systems <ufs-wm:build-wm>`.

Reviewer: ```suggestion
      Users not running on NOAA :term:`RDHPCS` will need to modify some additional files. See the :ref:`UFS WM regression testing documentation <ufs-wm:build-wm>` for more on how to do this.
```

Create a Working Directory
===========================

Create and navigate into a top-level working directory. This directory will be referred to as ``${TOP_DIR}`` throughout the tutorial.

.. code-block:: console

   mkdir wrk_dir
   cd wrk_dir

.. note:: 

   Users can optionally save their working directory in an environment variable. For example: 

   .. code-block:: console

      export TOP_DIR=/path/to/wrk_dir

.. _prepare-forecast:

Prepare Forecast Output
========================

#. The UPP requires forecast output as its input. In this example, we will use output created from a UFS Weather Model (WM) forecast run. This subsection is provided for ease of use, but users should refer to the :doc:`authoritative UFS WM documentation <ufs-wm:index>` and `UFS WM GitHub Discussions Q & A <https://github.com/ufs-community/ufs-weather-model/discussions/categories/q-a>`_ for questions:

   .. code-block:: console

      git clone --recursive https://github.com/ufs-community/ufs-weather-model.git
      cd ufs-weather-model/tests

#. Create a custom configuration file with the test or tests you want to run. 

   .. code-block:: console

      vi mytests.conf
      i

   .. note::

>       Users can check `rt.conf <https://github.com/ufs-community/ufs-weather-model/blob/develop/tests/rt.conf>`_ for a full list of tests and select the ones they want to run.
> 
>    .. COMMENT: Add info about files UPP expects? 

Reviewer: ```suggestion
      Users can check `rt.conf <https://github.com/ufs-community/ufs-weather-model/blob/develop/tests/rt.conf>`_ for a full list of tests and select the ones they want to run. The UPP expects ``atmf*``, ``sfcf*``, and ``GFSPRS*`` files for the GFS model, and it expects ``phyf*``, ``dynf*``, ``NATLEV*``, and ``PRSLEV*`` files for the LAM model, so users should run and/or adapt tests with these expected outputs.
```

#. Paste the following lines, which compile and run two sample cases (``control_p8`` and ``regional_control``), into ``mytests.conf``:

   .. code-block:: console

      COMPILE | atm_dyn32 | intel | -DAPP=ATM -DCCPP_SUITES=FV3_GFS_v16,FV3_GFS_v16_flake,FV3_GFS_v17_p8,FV3_GFS_v17_p8_rrtmgp,FV3_GFS_v15_thompson_mynn_lam3km,FV3_WoFS_v0,FV3_GFS_v17_p8_mynn,FV3_GFS_v17_p8_ugwpv1,FV3_GFS_v16_gfdlmpv3,FV3_GFS_v17_p8_ugwpv1_tempo -D32BIT=ON | | fv3 |
      RUN | control_p8                                        | - noaacloud                          | baseline |
      RUN | regional_control                                  |                                      | baseline |

#. Save and exit: Type ``esc`` and ``:wq``. Then hit ``Enter``/``Return``.

#. Edit each test's output frequency by modifying the test files under ``ufs-weather-model/tests/tests``. For example, to output every three hours for 24 hours:

   .. code-block:: console

      export OUTPUT_FH='3 -1'

   .. note:: 

      ``'3 -1'`` means every 3 hours. The ``-1`` indicates that it is not a list of hours. For example, someone could choose to output an arbitrary list of hours like ``export OUTPUT_FH='0 4 5 7 12 16 19'``. 

> 6. If you are not running on NOAA RDHPCS or do not have access to the ``stmp*`` disk space on those systems, alter the ``dprefix`` path for the machine you are running on in ``rt.sh`` to point to a directory (such as ``$TOP_DIR``) where you have write permissions. For example:

Reviewer: ```suggestion
6. If you are not running on NOAA :term:`RDHPCS` or do not have access to the ``stmp*`` disk space on those systems, alter the ``dprefix`` path for the machine you are running on in ``rt.sh`` to point to a directory (such as ``$TOP_DIR``) where you have write permissions. For example:
```

   .. code-block:: console

      dprefix=${dprefix:-"path/to/${TOP_DIR}"}

> 

Reviewer: @jdkublnick Could you add information here referring users to the WM documentation because if they're not on NOAA RDHPCS, there are other things they'll have to change, but the explanation is beyond the scope of this tutorial: https://ufs-weather-model.readthedocs.io/en/develop/BuildingAndRunning.html#other-systems

-----------------------------------

After modifying the XML postcontrol file, users must validate the file to ensure it was formatted successfully:

.. code-block:: console

   cd UPP/parm
   # For GFS/Global
   xmllint --noout --schema EMC_POST_CTRL_Schema.xsd gfs/postcntrl_gfs_f00_two.xml
   # For LAM/regional
   xmllint --noout --schema EMC_POST_CTRL_Schema.xsd rrfs/rrfs_postcntrl.xml

If the XML file is formatted correctly, it will print a message like:

.. code-block:: console

   gfs/postcntrl_gfs_f00_two.xml validates

.. _gen-text:

Generate the flat text file
----------------------------

If the control file requires any modifications, the user must convert the modified XML file to a flat text file. The command below will run the Perl program ``parm/PostXMLPreprocessor.pl`` to generate the ``postxconfig*`` flat file:

   .. code-block:: console

      cd UPP/parm
      # Global/GFS
      /usr/bin/perl PostXMLPreprocessor.pl gfs/postcntrl_gfs_f00_two.xml post_avblflds.xml gfs/postxconfig-NT-gfs-f00-two.txt
      # OR Regional/RRFS
      /usr/bin/perl PostXMLPreprocessor.pl rrfs/rrfs_postcntrl.xml post_avblflds.xml rrfs/postxconfig-NT-rrfs.txt

.. note::

>    ``PostXMLPreprocessor.pl`` must be run from the ``parm`` directory or it will produce an error. 

Reviewer: @jdkublnick Could you add this to the code snippets and place it in InputsOutputs line 253?

Reviewer: @jdkublnick Not sure this has been addressed? I think GitHub hid it... :-/

.. attention:: 

   On Ursa, the usual process for generating a new/updated ``postxconfig*.txt`` file is slightly different due to a missing XML module (see `Issue #1250 <https://github.com/NOAA-EMC/UPP/issues/1250>`_). The following workaround has been developed:

   .. code-block:: console

      wget https://raw.githubusercontent.com/wiki/NOAA-EMC/UPP/perl_venv_create.sh
      chmod 755 perl_venv_create.sh
      ./perl_venv_create.sh perl_venv
      source perl_venv/bin/activate
      cpanm XML::LibXML
      cd /path/to/UPP

>       cd parm

Reviewer: @jdkublnick Could you also turn this into a code snippet and place it also around line 253 of InputsOutputs?

Reviewer: @jdkublnick Not sure this has been addressed either? I think GitHub also hid it... :-/

For example in ``UPP/sorc/ncep_post.fd/SET_LVLSXML.f`` set:

.. code-block:: console

   ELSE  ! SPECIFY SIGO
      ASIGO1( 1)=   0.0000
      ASIGO1( 2)=   0.0500
      ASIGO1( 3)=   0.1000
      ASIGO1( 4)=   0.1500
      ASIGO1( 5)=   0.2000
      ASIGO1( 6)=   0.2500
      ASIGO1( 7)=   0.3000
      ASIGO1( 8)=   0.3500
      ASIGO1( 9)=   0.4000
      ASIGO1(10)=   0.4500
      ASIGO1(11)=   0.5000
      ASIGO1(12)=   0.5500
      ASIGO1(13)=   0.6000
      ASIGO1(14)=   0.6500
      ASIGO1(15)=   0.7000
      ASIGO1(16)=   0.7500
      ASIGO1(17)=   0.8000
      ASIGO1(18)=   0.8500
      ASIGO1(19)=   0.9000
      ASIGO1(20)=   0.9500
      ASIGO1(21)=   1.0000

.. _build-upp:

Build/Compile the UPP
----------------------

After modifying the source code, users must (re)build/(re)compile the UPP to ensure that changes take effect: 

.. code-block:: console

   cd UPP/tests
   ./compile_upp.sh

This will generate the UPP executable in the ``UPP/exec`` directory. 

.. _postprd:

Prepare the post-processing working directory
----------------------------------------------

#. Create a post-processing output directory:

   .. code-block:: console

>       cd $TOP_DIR

Reviewer: ```suggestion
      cd ${TOP_DIR}
```
For consistency :)

      mkdir postprd

   .. note::

>       This directory can be created anywhere, but default settings assume that is named ``postprd`` and created inside ``${TOP_DIR}``.

Reviewer: ```suggestion
      This directory can be created anywhere, but default settings assume that it is named ``postprd`` and created inside ``${TOP_DIR}``.
```

#. Download the UPP utility script for running standalone UPP and change the permissions:

   .. code-block:: console

      cd postprd
      wget https://raw.githubusercontent.com/wiki/NOAA-EMC/UPP/run_upp
      chmod 755 run_upp

> #. Modifying the script

Reviewer: ```suggestion
#. Modify the script:
```

   .. code-block:: console

      ...
      export TOP_DIR=/work2/noaa/epic/jsmith/hercules/test-sigma/
      export POSTPRD_DIR=${TOP_DIR}/postprd
      export UPP_HOME=${TOP_DIR}/UPP
      export POSTEXEC=${UPP_HOME}/exec
      export modelDataPath=${TOP_DIR}/control_p8_intel
      export txtCntrlFile=${UPP_HOME}/parm/gfs/postxconfig-NT-gfs-f00-two.txt
      export CRTMDIR=${UPP_HOME}/crtm/fix
      # Set date/time information
      export startdate=2021032206
      export fhr=00
      export lastfhr=06
      export incrementhr=03
      # Specify model ("GFS" or "LAM" in upper case)
      export model="GFS"
      # Set input format from model and ouput format from UPP
      export inFormat="netcdfpara"
      export outFormat="grib2"
      # Set run command: 
      # Single processor command example
      # export RUN_COMMAND="${POSTEXEC}/upp.x "

      #MPI sample command
      # "-n 4" can be changed to a different number of tasks. 
      export RUN_COMMAND="srun -A epic -n 4 ${POSTEXEC}/upp.x "

      # The number of subdomains in the x-direction (set to >=2 for 2d decomposition)
      export numx=1
      ...

   .. note::

>       The UPP expects ``atmf*``, ``sfcf*``, and ``GFSPRS*`` files for the GFS model, and it expects ``phyf*``, ``dynf*``, ``NATLEV*``, and ``PRSLEV*`` files for the LAM model. 

Reviewer: @jdkublnick This note could probably be turned into a code snippet because I think it appears in the run ch now, too (or should).

#. For regional (LAM) post-processing only:
   
   Regional/LAM post-processing requires satellite files, which can be downloaded from the UPP repository:

   .. code-block:: console

      mkdir crtm && cd crtm
      wget https://github.com/NOAA-EMC/UPP/releases/download/upp_v11.0.0/fix.tar.gz
      tar -xzf fix.tar.gz

   Make sure to adjust the ``CRTMDIR`` path in ``run_upp`` to point to the location of these files.

.. _run-runupp:

Run UPP
--------

#. Run the script:

   .. code-block:: console

>       ./run_upp -c <compiler> -m <hercules>

Reviewer: @jdkublnick See note in "RunningUPP" ch about putting this into a code snippet with info like "where ``<compiler>`` is "intel" or "intelllvm"..."

>    Users can add the ``-v`` option to see more output

Reviewer: ```suggestion
   Users can add the ``-v`` option to see more output.
```

   Check the ``upp.f{fhr}.out`` files to see if there were any errors when the script ran. 

#. Checking output:
   
   To view information about a file, use the wgrib2 utility. Users may need to load the wgrib2 module first (e.g., via ``module load wgrib2``) For example, to see summary information from ``GFSPRS.000``, run: 

   .. code-block:: console

      wgrib2 -s GFSPRS.000

>    For more detailed information, use the ``-v`` flag instead, but note that this is computationally intensive and may require allocating and using a compute node on some systems.

Reviewer: ```suggestion
   To view information about a file, use the *wgrib2* utility. Users may need to load the *wgrib2* module first (e.g., via ``module load wgrib2``). For example, to see summary information from ``GFSPRS.000``, run: 

   .. code-block:: console

      wgrib2 -s GFSPRS.000

   For more detailed information, use the ``-V`` flag instead, but note that this is computationally intensive and may require allocating and using a compute node on some systems.
```
`-V` flag must be capitalized, fyi.

      <fixed_sfc1_type>sigma_lvl</fixed_sfc1_type>
      <scale>4.0</scale>
      <level>0. 500. 1000. 1500. 2000. 2500. 3000. 3500. 4000. 4500. 5000. 5500. 6000. 6500. 7000. 7500. 8000. 8500. 9000. 9500. 10000.</level>
   </param>

.. _validate-xml:

Validate the XML postcontrol file
-----------------------------------

After modifying the XML postcontrol file, users must validate the file to ensure it was formatted successfully:

.. code-block:: console

   cd UPP/parm
   # For GFS/Global
   xmllint --noout --schema EMC_POST_CTRL_Schema.xsd gfs/postcntrl_gfs_f00_two.xml
   # For LAM/regional
   xmllint --noout --schema EMC_POST_CTRL_Schema.xsd rrfs/rrfs_postcntrl.xml

If the XML file is formatted correctly, it will print a message like:

.. code-block:: console

   gfs/postcntrl_gfs_f00_two.xml validates

.. _gen-text:

Generate the flat text file
----------------------------

If the control file requires any modifications, the user must convert the modified XML file to a flat text file. The command below will run the Perl program ``parm/PostXMLPreprocessor.pl`` to generate the ``postxconfig*`` flat file:

   .. code-block:: console

      cd UPP/parm
      # Global/GFS

>       /usr/bin/perl PostXMLPreprocessor.pl gfs/postcntrl_gfs_f00_two.xml
>       post_avblflds.xml gfs/postxconfig-NT-gfs-f00-two.txt
>       # OR Regional/RRFS
>       /usr/bin/perl PostXMLPreprocessor.pl rrfs/rrfs_postcntrl.xml post_avblflds.xml 
>       rrfs/postxconfig-NT-rrfs.txt

Reviewer: ```suggestion
      /usr/bin/perl PostXMLPreprocessor.pl gfs/postcntrl_gfs_f00_two.xml post_avblflds.xml gfs/postxconfig-NT-gfs-f00-two.txt
      # OR Regional/RRFS
      /usr/bin/perl PostXMLPreprocessor.pl rrfs/rrfs_postcntrl.xml post_avblflds.xml rrfs/postxconfig-NT-rrfs.txt
```
Just putting the commands fully on one line. Otherwise users could get tripped up if they copy-paste.

> .. _enabling-Output:

Reviewer: ```suggestion
.. _enabling-output:
```

********************************
User-Defined Sigma Level Output
********************************

This documentation describes the steps to generate model output at user-defined sigma levels using the Unified Post Processor (UPP). The instructions address the configuration of XML files and the resolution of potential issues encountered during this process.

Generating Sigma Level Output
-----------------------------

> UPP can output temperature, U, and V components on sigma surfaces. These correspond to indices 206, 208, and 209 in the :term:`GRIB2`` table. Follow these steps to configure your workflow for this output:

Reviewer: ```suggestion
UPP can output temperature, U, and V components on sigma surfaces. These correspond to indices 206, 208, and 209 in the :doc:`GRIB2 <UPP_GRIB2_Table_byID>` table. Follow these steps to configure your workflow for this output:
```

Reviewer: Could you actually link to the table? That's more important in terms of knowing what variables are possible in UPP.

**Configuring XML Files**

> 1. **Locate XML Configuration Files**:

Reviewer: It's best to replace the numbers with `#.` because then it autofills the numbers, and you don't have to worry about renumbering if steps are added or removed over time.

> 1. **Locate XML Configuration Files**:
>    - Use the XML files in the `UPP/parm` directory as references. Specifically, the `post_avblflds.xml` file contains all the fields that UPP can output.
> 
> 2. **Modify Control XML**:
>    - Copy the entries for temperature, U, and V (indices 206, 208, and 209) from `post_avblflds.xml` to the control XML file used in your workflow.
> 
> 3. **Validation**:
>    - Validate the control XML file against the `EMC_POST_CTRL_Schema.xsd` schema.
> 
> 4. **Flat Text File Generation**:
>    - Convert the control XML to a flat text file, ensuring that it is validated as part of the process. More information can be found in the :ref:`create_txt_file` documentation.

Reviewer: ```suggestion
An XML :ref:`control file <control-file>` determines what fields and levels UPP will output. 

#. **Locate the XML Control File**: Control files for various operational models are located in the ``UPP/parm`` directory. The ``post_avblflds.xml`` file contains all fields that the UPP can output. 

#. **Modify the Control XML**: Copy the entries for temperature, U, and V (indices 206, 208, and 209) from ``post_avblflds.xml`` to the control XML file used in your workflow. Be sure to include a ``<level></level>`` tag with the appropriate levels in the XML control file to define which sigma levels to output.

#. **Validate the Control File**: Validate the control XML file against the ``EMC_POST_CTRL_Schema.xsd`` schema according to :ref:`these instructions <create_txt_file>`.

#. **Flat Text File Generation**: Convert the control XML to a flat text file based on the information in the :ref:`create_txt_file` documentation.
```

**Sigma Levels**

> By default, only certain sigma levels are outputted. These levels are in the source file `SET_LVLSXML.f` and can be found between lines 342 and 363. Users must:

Reviewer: ```suggestion
By default, only certain sigma levels are outputted. These levels are defined in `SET_LVLSXML.f <https://github.com/NOAA-EMC/UPP/blob/develop/sorc/ncep_post.fd/SET_LVLSXML.f>`_ using the ``ASIGO1`` array. Users must review these levels in the ``SET_LVLSXML.f`` file to confirm compatibility with their requirements.
```
Best not to indicate specific lines of code because they can change.

> - **Verify Default Levels**:
>   Review these levels in the `SET_LVLSXML.f` source file to confirm their compatibility with your requirements.
> 
> - **Add <level></level> Tags**:
>   Include a `<level></level>` tag in the XML configuration to explicitly define the sigma levels to be output.

Reviewer: ```suggestion
```
The level tags pertain to the control file, not the SET_LVLSXML.f, so I moved that info up and combined the remaining bullet point into the sentence above.

Required Source Code Modifications
----------------------------------

> If the default sigma levels are insufficient, you must:
> 
> 1. **Modify `SET_LVLSXML.f`**:
>    - Change the entries for `ASIGO1` to include your desired sigma levels.
> 
> 2. **Recompile UPP**:
>    - Recompile the UPP source code after making the changes.
> 
> 3. **Update XML Tags**:
>    - Ensure that the XML tags explicitly reference the modified sigma levels.

Reviewer: ```suggestion
If the default sigma levels are insufficient, users must:

#. **Modify** ``SET_LVLSXML.f``: Change the entries for ``ASIGO1`` to include your desired sigma levels.

#. **Recompile UPP**: Recompile the UPP source code after making the changes.

#. **Update XML Tags**: Ensure that the XML control file explicitly references the modified sigma levels in the ``<level></level>`` tag.
```

> Important Notes
> ---------------
> 
> - Adding `<level></level>` tags manually allows UPP to compile and run successfully, but it will not produce any output unless the levels are also modified in `SET_LVLSXML.f` and properly referenced.
> 
> - Fortran I/O errors may occur if the sigma levels in the control file are not synchronized with the source code configuration.
> 
> - Comprehensive documentation on control files can be found in the :ref:`control-file` documentation, which provides details about XML file formatting and flat file generation.
> 
> By following these steps, users can successfully configure UPP to generate model output on user-defined sigma levels while avoiding common pitfalls.

Reviewer: ```suggestion
.. hint::

   - Adding ``<level></level>`` tags manually allows UPP to compile and run successfully, but it will not produce any output unless the levels are also modified in ``SET_LVLSXML.f`` and properly referenced.

   - Fortran I/O errors may occur if the sigma levels in the control file are not synchronized with the source code configuration.

   - Comprehensive documentation on control files can be found in the :ref:`control-file` documentation, which provides details about XML file formatting and flat file generation.
```
